File: example/export/go/protocol.go
package protocol

import (
	"bytes"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"math"
)

const (
	protocolMaxByteLen = 16 << 20 // 单段字符串/嵌套二进制最大 16MiB
	protocolMaxCount   = 1 << 20  // 数组/map 最大元素数
)

type ProtocolLogin struct {
	A int8      `json:"a"`
	B uint8     `json:"b"`
	C int16     `json:"c"`
	D uint16    `json:"d"`
	E int32     `json:"e"`
	F uint32    `json:"f"`
	G int64     `json:"g"`
	H uint64    `json:"h"`
	I float32   `json:"i"`
	J float64   `json:"j"`
	K bool      `json:"k"`
	L byte      `json:"l"`
	M string    `json:"m"`
	N []int8    `json:"n"`
	O []string  `json:"o"`
	Q Account   `json:"q"`
	R []Account `json:"r"`
}

func (this *ProtocolLogin) Decode(data []byte) error {
	if len(data) == 0 {
		return nil
	}
	if data[0] == '{' {
		if err := json.Unmarshal(data, this); err != nil {
			return fmt.Errorf("json decode: %w", err)
		}
		return nil
	}
	var pointer int
	if pointer+1 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.A = int8(data[pointer])
	pointer += 1
	if pointer+1 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.B = uint8(data[pointer])
	pointer += 1
	if pointer+2 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.C = int16(binary.LittleEndian.Uint16(data[pointer : pointer+2]))
	pointer += 2
	if pointer+2 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.D = binary.LittleEndian.Uint16(data[pointer : pointer+2])
	pointer += 2
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.E = int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
	pointer += 4
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.F = binary.LittleEndian.Uint32(data[pointer : pointer+4])
	pointer += 4
	if pointer+8 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.G = int64(binary.LittleEndian.Uint64(data[pointer : pointer+8]))
	pointer += 8
	if pointer+8 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.H = binary.LittleEndian.Uint64(data[pointer : pointer+8])
	pointer += 8
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.I = math.Float32frombits(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
	pointer += 4
	if pointer+8 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.J = math.Float64frombits(binary.LittleEndian.Uint64(data[pointer : pointer+8]))
	pointer += 8
	if pointer+1 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.K = data[pointer] != 0
	pointer += 1
	if pointer+1 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.L = byte(data[pointer])
	pointer += 1
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	{
		length := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
		pointer += 4
		if length < 0 {
			return fmt.Errorf("protocol: negative length")
		}
		if int(length) > protocolMaxByteLen {
			return fmt.Errorf("protocol: length %d exceeds limit %d", length, protocolMaxByteLen)
		}
		if pointer+int(length) > len(data) {
			return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
		}
		this.M = string(data[pointer : pointer+int(length)])
		pointer += int(length)
	}
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	{
		count := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
		pointer += 4
		if count < 0 {
			return fmt.Errorf("protocol: negative count")
		}
		if int(count) > protocolMaxCount {
			return fmt.Errorf("protocol: count %d exceeds limit %d", count, protocolMaxCount)
		}
		this.N = this.N[:0]
		for i := 0; i < int(count); i++ {
			if pointer+1 > len(data) {
				return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
			}
			this.N = append(this.N, int8(data[pointer]))
			pointer += 1
		}
	}
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	{
		count := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
		pointer += 4
		if count < 0 {
			return fmt.Errorf("protocol: negative count")
		}
		if int(count) > protocolMaxCount {
			return fmt.Errorf("protocol: count %d exceeds limit %d", count, protocolMaxCount)
		}
		this.O = this.O[:0]
		for i := 0; i < int(count); i++ {
			if pointer+4 > len(data) {
				return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
			}
			{
				length := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
				pointer += 4
				if length < 0 {
					return fmt.Errorf("protocol: negative length")
				}
				if int(length) > protocolMaxByteLen {
					return fmt.Errorf("protocol: length %d exceeds limit %d", length, protocolMaxByteLen)
				}
				if pointer+int(length) > len(data) {
					return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
				}
				this.O = append(this.O, string(data[pointer:pointer+int(length)]))
				pointer += int(length)
			}
		}
	}
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	{
		length := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
		pointer += 4
		if length < 0 {
			return fmt.Errorf("protocol: negative length")
		}
		if int(length) > protocolMaxByteLen {
			return fmt.Errorf("protocol: length %d exceeds limit %d", length, protocolMaxByteLen)
		}
		if pointer+int(length) > len(data) {
			return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
		}
		this.Q = Account{}
		if err := this.Q.Decode(data[pointer : pointer+int(length)]); err != nil {
			return err
		}
		pointer += int(length)
	}
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	{
		count := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
		pointer += 4
		if count < 0 {
			return fmt.Errorf("protocol: negative count")
		}
		if int(count) > protocolMaxCount {
			return fmt.Errorf("protocol: count %d exceeds limit %d", count, protocolMaxCount)
		}
		this.R = this.R[:0]
		for i := 0; i < int(count); i++ {
			if pointer+4 > len(data) {
				return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
			}
			{
				length := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
				pointer += 4
				if length < 0 {
					return fmt.Errorf("protocol: negative length")
				}
				if int(length) > protocolMaxByteLen {
					return fmt.Errorf("protocol: length %d exceeds limit %d", length, protocolMaxByteLen)
				}
				if pointer+int(length) > len(data) {
					return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
				}
				var value = Account{}
				if err := value.Decode(data[pointer : pointer+int(length)]); err != nil {
					return err
				}
				this.R = append(this.R, value)
				pointer += int(length)
			}
		}
	}
	return nil
}

func (this *ProtocolLogin) EncodeBinary() (data []byte) {
	var buffer = bytes.NewBuffer([]byte{})
	binary.Write(buffer, binary.LittleEndian, this.A)
	binary.Write(buffer, binary.LittleEndian, this.B)
	binary.Write(buffer, binary.LittleEndian, this.C)
	binary.Write(buffer, binary.LittleEndian, this.D)
	binary.Write(buffer, binary.LittleEndian, this.E)
	binary.Write(buffer, binary.LittleEndian, this.F)
	binary.Write(buffer, binary.LittleEndian, this.G)
	binary.Write(buffer, binary.LittleEndian, this.H)
	binary.Write(buffer, binary.LittleEndian, this.I)
	binary.Write(buffer, binary.LittleEndian, this.J)
	binary.Write(buffer, binary.LittleEndian, this.K)
	binary.Write(buffer, binary.LittleEndian, this.L)
	{
		var b = []byte(this.M)
		var l = int32(len(b))
		binary.Write(buffer, binary.LittleEndian, l)
		binary.Write(buffer, binary.LittleEndian, b)
	}
	{
		var count = int32(len(this.N))
		binary.Write(buffer, binary.LittleEndian, count)
		for _, item := range this.N {
			binary.Write(buffer, binary.LittleEndian, item)
		}
	}
	{
		var count = int32(len(this.O))
		binary.Write(buffer, binary.LittleEndian, count)
		for _, item := range this.O {
			var b = []byte(item)
			var l = int32(len(b))
			binary.Write(buffer, binary.LittleEndian, l)
			binary.Write(buffer, binary.LittleEndian, b)
		}
	}
	{
		var b = (this.Q).EncodeBinary()
		var l = int32(len(b))
		binary.Write(buffer, binary.LittleEndian, l)
		binary.Write(buffer, binary.LittleEndian, b)
	}
	{
		var count = int32(len(this.R))
		binary.Write(buffer, binary.LittleEndian, count)
		for _, item := range this.R {
			var b = item.EncodeBinary()
			var l = int32(len(b))
			binary.Write(buffer, binary.LittleEndian, l)
			binary.Write(buffer, binary.LittleEndian, b)
		}
	}
	data = buffer.Bytes()
	return
}

func (this *ProtocolLogin) EncodeJson() (data []byte) {
	data, _ = json.Marshal(this)
	return
}

type Account struct {
	Nickname string `json:"nickname"`
	Password string `json:"password"`
}

func (this *Account) Decode(data []byte) error {
	if len(data) == 0 {
		return nil
	}
	if data[0] == '{' {
		if err := json.Unmarshal(data, this); err != nil {
			return fmt.Errorf("json decode: %w", err)
		}
		return nil
	}
	var pointer int
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	{
		length := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
		pointer += 4
		if length < 0 {
			return fmt.Errorf("protocol: negative length")
		}
		if int(length) > protocolMaxByteLen {
			return fmt.Errorf("protocol: length %d exceeds limit %d", length, protocolMaxByteLen)
		}
		if pointer+int(length) > len(data) {
			return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
		}
		this.Nickname = string(data[pointer : pointer+int(length)])
		pointer += int(length)
	}
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	{
		length := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
		pointer += 4
		if length < 0 {
			return fmt.Errorf("protocol: negative length")
		}
		if int(length) > protocolMaxByteLen {
			return fmt.Errorf("protocol: length %d exceeds limit %d", length, protocolMaxByteLen)
		}
		if pointer+int(length) > len(data) {
			return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
		}
		this.Password = string(data[pointer : pointer+int(length)])
		pointer += int(length)
	}
	return nil
}

func (this *Account) EncodeBinary() (data []byte) {
	var buffer = bytes.NewBuffer([]byte{})
	{
		var b = []byte(this.Nickname)
		var l = int32(len(b))
		binary.Write(buffer, binary.LittleEndian, l)
		binary.Write(buffer, binary.LittleEndian, b)
	}
	{
		var b = []byte(this.Password)
		var l = int32(len(b))
		binary.Write(buffer, binary.LittleEndian, l)
		binary.Write(buffer, binary.LittleEndian, b)
	}
	data = buffer.Bytes()
	return
}

func (this *Account) EncodeJson() (data []byte) {
	data, _ = json.Marshal(this)
	return
}

type ProtocolServerLogin struct {
	ID    string `json:"id"`
	Token string `json:"token"`
}

func (this *ProtocolServerLogin) Decode(data []byte) error {
	if len(data) == 0 {
		return nil
	}
	if data[0] == '{' {
		if err := json.Unmarshal(data, this); err != nil {
			return fmt.Errorf("json decode: %w", err)
		}
		return nil
	}
	var pointer int
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	{
		length := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
		pointer += 4
		if length < 0 {
			return fmt.Errorf("protocol: negative length")
		}
		if int(length) > protocolMaxByteLen {
			return fmt.Errorf("protocol: length %d exceeds limit %d", length, protocolMaxByteLen)
		}
		if pointer+int(length) > len(data) {
			return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
		}
		this.ID = string(data[pointer : pointer+int(length)])
		pointer += int(length)
	}
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	{
		length := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
		pointer += 4
		if length < 0 {
			return fmt.Errorf("protocol: negative length")
		}
		if int(length) > protocolMaxByteLen {
			return fmt.Errorf("protocol: length %d exceeds limit %d", length, protocolMaxByteLen)
		}
		if pointer+int(length) > len(data) {
			return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
		}
		this.Token = string(data[pointer : pointer+int(length)])
		pointer += int(length)
	}
	return nil
}

func (this *ProtocolServerLogin) EncodeBinary() (data []byte) {
	var buffer = bytes.NewBuffer([]byte{})
	{
		var b = []byte(this.ID)
		var l = int32(len(b))
		binary.Write(buffer, binary.LittleEndian, l)
		binary.Write(buffer, binary.LittleEndian, b)
	}
	{
		var b = []byte(this.Token)
		var l = int32(len(b))
		binary.Write(buffer, binary.LittleEndian, l)
		binary.Write(buffer, binary.LittleEndian, b)
	}
	data = buffer.Bytes()
	return
}

func (this *ProtocolServerLogin) EncodeJson() (data []byte) {
	data, _ = json.Marshal(this)
	return
}

type ProtocolServerState struct {
	State uint16 `json:"state"`
	Value string `json:"value"`
}

func (this *ProtocolServerState) Decode(data []byte) error {
	if len(data) == 0 {
		return nil
	}
	if data[0] == '{' {
		if err := json.Unmarshal(data, this); err != nil {
			return fmt.Errorf("json decode: %w", err)
		}
		return nil
	}
	var pointer int
	if pointer+2 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.State = binary.LittleEndian.Uint16(data[pointer : pointer+2])
	pointer += 2
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	{
		length := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
		pointer += 4
		if length < 0 {
			return fmt.Errorf("protocol: negative length")
		}
		if int(length) > protocolMaxByteLen {
			return fmt.Errorf("protocol: length %d exceeds limit %d", length, protocolMaxByteLen)
		}
		if pointer+int(length) > len(data) {
			return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
		}
		this.Value = string(data[pointer : pointer+int(length)])
		pointer += int(length)
	}
	return nil
}

func (this *ProtocolServerState) EncodeBinary() (data []byte) {
	var buffer = bytes.NewBuffer([]byte{})
	binary.Write(buffer, binary.LittleEndian, this.State)
	{
		var b = []byte(this.Value)
		var l = int32(len(b))
		binary.Write(buffer, binary.LittleEndian, l)
		binary.Write(buffer, binary.LittleEndian, b)
	}
	data = buffer.Bytes()
	return
}

func (this *ProtocolServerState) EncodeJson() (data []byte) {
	data, _ = json.Marshal(this)
	return
}

type ProtocolServerCommand struct {
	Command uint16 `json:"command"`
	Value   string `json:"value"`
}

func (this *ProtocolServerCommand) Decode(data []byte) error {
	if len(data) == 0 {
		return nil
	}
	if data[0] == '{' {
		if err := json.Unmarshal(data, this); err != nil {
			return fmt.Errorf("json decode: %w", err)
		}
		return nil
	}
	var pointer int
	if pointer+2 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	this.Command = binary.LittleEndian.Uint16(data[pointer : pointer+2])
	pointer += 2
	if pointer+4 > len(data) {
		return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
	}
	{
		length := int32(binary.LittleEndian.Uint32(data[pointer : pointer+4]))
		pointer += 4
		if length < 0 {
			return fmt.Errorf("protocol: negative length")
		}
		if int(length) > protocolMaxByteLen {
			return fmt.Errorf("protocol: length %d exceeds limit %d", length, protocolMaxByteLen)
		}
		if pointer+int(length) > len(data) {
			return fmt.Errorf("protocol: unexpected end of data at offset %d", pointer)
		}
		this.Value = string(data[pointer : pointer+int(length)])
		pointer += int(length)
	}
	return nil
}

func (this *ProtocolServerCommand) EncodeBinary() (data []byte) {
	var buffer = bytes.NewBuffer([]byte{})
	binary.Write(buffer, binary.LittleEndian, this.Command)
	{
		var b = []byte(this.Value)
		var l = int32(len(b))
		binary.Write(buffer, binary.LittleEndian, l)
		binary.Write(buffer, binary.LittleEndian, b)
	}
	data = buffer.Bytes()
	return
}

func (this *ProtocolServerCommand) EncodeJson() (data []byte) {
	data, _ = json.Marshal(this)
	return
}

File: example/export/csharp/Protocol.cs
using System;
using System.IO;
using System.Text;
using System.Text.Json;

namespace protocol
{
public class ProtocolLogin
{
public sbyte a;
public byte b;
public Int16 c;
public UInt16 d;
public Int32 e;
public UInt32 f;
public Int64 g;
public UInt64 h;
public float i;
public double j;
public bool k;
public byte l;
public string m;
public sbyte[] n;
public string[] o;
public Account q;
public Account[] r;
    public ProtocolLogin()
    {
a = 0;
b = 0;
c = 0;
d = 0;
e = 0;
f = 0;
g = 0;
h = 0;
i = 0;
j = 0;
k = false;
l = 0;
m = "";
n = new sbyte[]{};
o = new string[]{};
q = new Account();
r = new Account[]{};
    }
    public void Decoder(byte[] data)
    {
        if (data == null || data.Length == 0) return;
        if (Encoding.Default.GetString(new byte[] { data[0] }) == "{")
        {
            var protocol = JsonSerializer.Deserialize<ProtocolLogin>(Encoding.Default.GetString(data));
a = protocol.a;
b = protocol.b;
c = protocol.c;
d = protocol.d;
e = protocol.e;
f = protocol.f;
g = protocol.g;
h = protocol.h;
i = protocol.i;
j = protocol.j;
k = protocol.k;
l = protocol.l;
m = protocol.m;
n = protocol.n;
o = protocol.o;
q = protocol.q;
r = protocol.r;
        }
    	else
    	{
            MemoryStream steam = new MemoryStream(data);
            BinaryReader reader = new BinaryReader(steam);
a = reader.ReadSByte();
b = reader.ReadByte();
c = reader.ReadInt16();
d = reader.ReadUInt16();
e = reader.ReadInt32();
f = reader.ReadUInt32();
g = reader.ReadInt64();
h = reader.ReadUInt64();
i = reader.ReadSingle();
j = reader.ReadDouble();
k = reader.ReadBoolean();
l = reader.ReadByte();
            var mLenght = reader.ReadInt32();
            var mReader = new BinaryReader(new MemoryStream(reader.ReadBytes(mLenght)));
            this.m = mReader.ReadString();
            var nCount = reader.ReadInt32();
            this.n = new sbyte[nCount];
            for (var i = 0; i < nCount; i++)
            {
this.n[i] = reader.ReadSByte();
            }
            var oCount = reader.ReadInt32();
            this.o = new string[oCount];
            for (var i = 0; i < oCount; i++)
            {
                var l = reader.ReadInt32();
                var r = new BinaryReader(new MemoryStream(reader.ReadBytes(l)));
                this.o[i] = r.ReadString();
            }
            var qLenght = reader.ReadInt32();
            var qByte = reader.ReadBytes(qLenght);
            this.q = new Account();
            this.q.Decoder(qByte);
            var rCount = reader.ReadInt32();
            this.r = new Account[rCount];
            for (var i = 0; i < rCount; i++)
            {
                var l = reader.ReadInt32();
                var d = reader.ReadBytes(l);
                this.r[i] = new Account();
                this.r[i].Decoder(d);
            }
    	}
    }
    public byte[] EncodeBinary()
    {
        MemoryStream steam = new MemoryStream();
        BinaryWriter writer = new BinaryWriter(steam);
writer.Write(this.a);
writer.Write(this.b);
writer.Write(this.c);
writer.Write(this.d);
writer.Write(this.e);
writer.Write(this.f);
writer.Write(this.g);
writer.Write(this.h);
writer.Write(this.i);
writer.Write(this.j);
writer.Write(this.k);
writer.Write(this.l);
        var mWriter = new BinaryWriter(new MemoryStream());
        mWriter.Write(this.m);
        writer.Write((Int32)mWriter.BaseStream.Length);
        writer.Write(this.m);
writer.Write((Int32)this.n.Length);
for (var i = 0; i < this.n.Length; i++)
{
            writer.Write(this.n[i]);
}
writer.Write((Int32)this.o.Length);
for (var i = 0; i < this.o.Length; i++)
{
            var w = new BinaryWriter(new MemoryStream());
            w.Write(this.o[i]);
            writer.Write((Int32)w.BaseStream.Length);
            writer.Write(this.o[i]);
}
        var qByteArray = this.q.EncodeBinary();
        writer.Write((Int32)qByteArray.Length);
        writer.Write(qByteArray);
writer.Write((Int32)this.r.Length);
for (var i = 0; i < this.r.Length; i++)
{
            var array = this.r[i].EncodeBinary();
            writer.Write((Int32)array.Length);
            writer.Write(array);
}
    return steam.ToArray();
    }
    public byte[] EncodeJson()
    {
        return Encoding.Default.GetBytes(JsonSerializer.Serialize(this));
    }
}
public class Account
{
public string nickname;
public string password;
    public Account()
    {
nickname = "";
password = "";
    }
    public void Decoder(byte[] data)
    {
        if (data == null || data.Length == 0) return;
        if (Encoding.Default.GetString(new byte[] { data[0] }) == "{")
        {
            var protocol = JsonSerializer.Deserialize<Account>(Encoding.Default.GetString(data));
nickname = protocol.nickname;
password = protocol.password;
        }
    	else
    	{
            MemoryStream steam = new MemoryStream(data);
            BinaryReader reader = new BinaryReader(steam);
            var nicknameLenght = reader.ReadInt32();
            var nicknameReader = new BinaryReader(new MemoryStream(reader.ReadBytes(nicknameLenght)));
            this.nickname = nicknameReader.ReadString();
            var passwordLenght = reader.ReadInt32();
            var passwordReader = new BinaryReader(new MemoryStream(reader.ReadBytes(passwordLenght)));
            this.password = passwordReader.ReadString();
    	}
    }
    public byte[] EncodeBinary()
    {
        MemoryStream steam = new MemoryStream();
        BinaryWriter writer = new BinaryWriter(steam);
        var nicknameWriter = new BinaryWriter(new MemoryStream());
        nicknameWriter.Write(this.nickname);
        writer.Write((Int32)nicknameWriter.BaseStream.Length);
        writer.Write(this.nickname);
        var passwordWriter = new BinaryWriter(new MemoryStream());
        passwordWriter.Write(this.password);
        writer.Write((Int32)passwordWriter.BaseStream.Length);
        writer.Write(this.password);
    return steam.ToArray();
    }
    public byte[] EncodeJson()
    {
        return Encoding.Default.GetBytes(JsonSerializer.Serialize(this));
    }
}
public class ProtocolServerLogin
{
public string id;
public string token;
    public ProtocolServerLogin()
    {
id = "";
token = "";
    }
    public void Decoder(byte[] data)
    {
        if (data == null || data.Length == 0) return;
        if (Encoding.Default.GetString(new byte[] { data[0] }) == "{")
        {
            var protocol = JsonSerializer.Deserialize<ProtocolServerLogin>(Encoding.Default.GetString(data));
id = protocol.id;
token = protocol.token;
        }
    	else
    	{
            MemoryStream steam = new MemoryStream(data);
            BinaryReader reader = new BinaryReader(steam);
            var idLenght = reader.ReadInt32();
            var idReader = new BinaryReader(new MemoryStream(reader.ReadBytes(idLenght)));
            this.id = idReader.ReadString();
            var tokenLenght = reader.ReadInt32();
            var tokenReader = new BinaryReader(new MemoryStream(reader.ReadBytes(tokenLenght)));
            this.token = tokenReader.ReadString();
    	}
    }
    public byte[] EncodeBinary()
    {
        MemoryStream steam = new MemoryStream();
        BinaryWriter writer = new BinaryWriter(steam);
        var idWriter = new BinaryWriter(new MemoryStream());
        idWriter.Write(this.id);
        writer.Write((Int32)idWriter.BaseStream.Length);
        writer.Write(this.id);
        var tokenWriter = new BinaryWriter(new MemoryStream());
        tokenWriter.Write(this.token);
        writer.Write((Int32)tokenWriter.BaseStream.Length);
        writer.Write(this.token);
    return steam.ToArray();
    }
    public byte[] EncodeJson()
    {
        return Encoding.Default.GetBytes(JsonSerializer.Serialize(this));
    }
}
public class ProtocolServerState
{
public UInt16 state;
public string value;
    public ProtocolServerState()
    {
state = 0;
value = "";
    }
    public void Decoder(byte[] data)
    {
        if (data == null || data.Length == 0) return;
        if (Encoding.Default.GetString(new byte[] { data[0] }) == "{")
        {
            var protocol = JsonSerializer.Deserialize<ProtocolServerState>(Encoding.Default.GetString(data));
state = protocol.state;
value = protocol.value;
        }
    	else
    	{
            MemoryStream steam = new MemoryStream(data);
            BinaryReader reader = new BinaryReader(steam);
state = reader.ReadUInt16();
            var valueLenght = reader.ReadInt32();
            var valueReader = new BinaryReader(new MemoryStream(reader.ReadBytes(valueLenght)));
            this.value = valueReader.ReadString();
    	}
    }
    public byte[] EncodeBinary()
    {
        MemoryStream steam = new MemoryStream();
        BinaryWriter writer = new BinaryWriter(steam);
writer.Write(this.state);
        var valueWriter = new BinaryWriter(new MemoryStream());
        valueWriter.Write(this.value);
        writer.Write((Int32)valueWriter.BaseStream.Length);
        writer.Write(this.value);
    return steam.ToArray();
    }
    public byte[] EncodeJson()
    {
        return Encoding.Default.GetBytes(JsonSerializer.Serialize(this));
    }
}
public class ProtocolServerCommand
{
public UInt16 command;
public string value;
    public ProtocolServerCommand()
    {
command = 0;
value = "";
    }
    public void Decoder(byte[] data)
    {
        if (data == null || data.Length == 0) return;
        if (Encoding.Default.GetString(new byte[] { data[0] }) == "{")
        {
            var protocol = JsonSerializer.Deserialize<ProtocolServerCommand>(Encoding.Default.GetString(data));
command = protocol.command;
value = protocol.value;
        }
    	else
    	{
            MemoryStream steam = new MemoryStream(data);
            BinaryReader reader = new BinaryReader(steam);
command = reader.ReadUInt16();
            var valueLenght = reader.ReadInt32();
            var valueReader = new BinaryReader(new MemoryStream(reader.ReadBytes(valueLenght)));
            this.value = valueReader.ReadString();
    	}
    }
    public byte[] EncodeBinary()
    {
        MemoryStream steam = new MemoryStream();
        BinaryWriter writer = new BinaryWriter(steam);
writer.Write(this.command);
        var valueWriter = new BinaryWriter(new MemoryStream());
        valueWriter.Write(this.value);
        writer.Write((Int32)valueWriter.BaseStream.Length);
        writer.Write(this.value);
    return steam.ToArray();
    }
    public byte[] EncodeJson()
    {
        return Encoding.Default.GetBytes(JsonSerializer.Serialize(this));
    }
}
}
